packages/@lwc/engine-core/src/framework/modules/static-parts.ts

```diff
@@ -82,16 +93,13 @@ export function traverseAndSetElements(
         } else {
             let sibling: Element | Text | null;
             while (isNull((sibling = nextSibling(node)))) {
-                if (process.env.NODE_ENV !== 'production') {
-                    // We should never traverse up to the root. We should exit early due to numFoundParts === numParts.
-                    assert.isFalse(
-                        node === root,
-                        `Reached the root without finding all parts. Found ${numFoundParts}, needed ${numParts}.`
-                    );
-                }
+                // we never want to walk up from the root
+                assertNotRoot(node);
                 // walk up
                 node = getParentNode(node);
             }
+            // we never want to walk right from the root
+            assertNotRoot(node);
```

Comment:
Note I have a similar PR forthcoming for `@lwc/synthetic-shadow` that demonstrates that it is possible to mistakenly traverse right from the root in this algorithm. But in these tests this case is not possible to hit.